metaflow/datastore/task_datastore.py

```diff
@@ -511,38 +508,50 @@ def persist(self, flow):
         """
         Persist any new artifacts that were produced when running flow
 
+        NOTE: This is a DESTRUCTIVE operation that deletes artifacts from
+        the given flow to conserve memory. Don't rely on artifact attributes
+        of the flow object after calling this function.
+
         Parameters
         ----------
         flow : FlowSpec
             Flow to persist
         """
-        def serializable_attributes():
-            for var in dir(flow):
-                if var.startswith('__') or var in flow._EPHEMERAL:
-                    continue
-                # Skip over properties of the class (Parameters)
-                if hasattr(flow.__class__, var) and \
-                        isinstance(getattr(flow.__class__, var), property):
-                    continue
-                val = getattr(flow, var)
-                if not (isinstance(val, MethodType) or
-                        isinstance(val, FunctionType) or
-                        isinstance(val, Parameter)):
-                    yield var, val, False
-
-        # initialize with old values...
+
         if flow._datastore:
             self._objects.update(flow._datastore._objects)
             self._info.update(flow._datastore._info)
 
-        # ...overwrite with new
-        to_save = {}
-        force_v4_dict = {}
-        for var, obj, force_v4 in serializable_attributes():
-            to_save[var] = obj
-            if force_v4:
-                force_v4_dict[var] = True
-        self.save_artifacts(to_save, force_v4_dict)
+        # we create a list of valid_artifacts in advance, outside of
+        # artifacts_iter so we can provide a len_hint below
+        valid_artifacts = []
+        for var in dir(flow):
+            if var.startswith('__') or var in flow._EPHEMERAL:
+                continue
+            # Skip over properties of the class (Parameters)
+            if hasattr(flow.__class__, var) and \
+                    isinstance(getattr(flow.__class__, var), property):
+                continue
+
+            val = getattr(flow, var)
+            if not (isinstance(val, MethodType) or
+                    isinstance(val, FunctionType) or
+                    isinstance(val, Parameter)):
+                valid_artifacts.append((var, val))
+
+        def artifacts_iter():
+            # we consume the valid_artifacts list destructively to
+            # make sure we don't keep references to artifacts. We
+            # want to avoid keeping original artifacts and encoded
+            # artifacts in memory simultaneously
+            while valid_artifacts:
+                var, val = valid_artifacts.pop()
+                if not var.startswith('_') and var != 'name':
```

Comment:
Maybe put a small comment here that we exclude names starting with '_' and 'name' because they are used internally (I was lazy and didn't list all of them explicitly).